### DISCUSSION TITLE
Can ui.refreshable support better type hints?

### DISCUSSION THREAD

**Discussion by CrystalWindSnake** [zauberzeug/nicegui#1951] [Ideas / Feature Requests]
Currently, the function object wrapped by `ui.refreshable` loses the parameter types . I found a solution in `ParamSpec` from `typing_extensions`, not sure if it's feasible.

**Comment by falkoschindler** [comment]
Interesting! I'm always looking for ways to improve type hints in NiceGUI.

If I understand correctly, you mean the missing types in a case like this:

<img width="517" alt="Screenshot 2023-11-03 at 16 17 22" src="https://github.com/zauberzeug/nicegui/assets/5767091/54848559-1904-423b-afea-d67e4c130cfb">

Can you elaborate on how we can use `ParamSpec`?

**Reply by CrystalWindSnake** [reply]
Like this, it seems to work. But I'm not sure if the return type of the `__call__` function is set up properly

```python

from typing_extensions import ParamSpec 
_T = TypeVar("_T")
_P = ParamSpec('_P')


class refreshable(Generic[_P, _T]):
       def __init__(self, func: Callable[_P, _T]) -> None:
            ...

       def __call__(self, *args: _P.args, **kwargs: _P.kwargs) -> Union[Optional[_T], Awaitable]:
             ...

       def refresh(self, *args: _P.args, **kwargs: _P.kwargs) -> Optional[_T]:
             ...
     
```

![image](https://github.com/zauberzeug/nicegui/assets/36034954/027c3743-8ee1-4240-bfa5-903c0c805e93)

![image](https://github.com/zauberzeug/nicegui/assets/36034954/0814875f-5637-455d-822c-b49c76aee570)

**Reply by falkoschindler** [reply]
Awesome! This was a great starting point for PR #1969. 🙂
